### ISSUE TITLE
Enhancement: include leadingComments and trailingComments in types package

### ISSUE THREAD

**Issue by bschlenk** [typescript-eslint/typescript-eslint#5195]
### Before You File a Proposal Please Confirm You Have Done The Following...

- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22enhancement%3A+plugin+rule+option%22) and found none that match my proposal.
- [X] I have searched the [current rule list](https://typescript-eslint.io/rules/#supported-rules) and found no rules that match my proposal.
- [X] I have [read the FAQ](https://typescript-eslint.io/docs/linting/troubleshooting) and my problem is not listed.

### Relevant Package

typescript-estree

### My proposal is suitable for this project

- [X] I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal).

### Description

Hello! I was trying to write an eslint rule to enforce some things with jsdoc style comments, and I noticed that the typescript types do not include `leadingComments` / `trailingComments` on any of the nodes. 

I tried diving in to see if it would be easy to add, but the types look like they're generated from something and I didn't dive in much further. I'd be happy to help out with a PR if I could get some direction as to where this would be added. Thanks!

### Fail

```typescript
N/A
```

### Pass

```typescript
N/A
```

### Additional Info

The relevant package dropdown didn't include the `types` package, so I semi-randomly chose typescript-estree. Sorry if this makes triage difficult!

**Comment by bradzacher** [comment]
`leadingComments` and `trailingComments` are not standard ESTree AST properties.
We will not be adding these properties.

You should use ESLint APIs to fetch comments, available on `context.getSourceCode()`
https://github.com/typescript-eslint/typescript-eslint/blob/2a345b40508c61917b0488f688c94407f79cb7f2/packages/utils/src/ts-eslint/SourceCode.ts#L8-L39

**Comment by bschlenk** [comment]
Ahh, thanks! That works just as well.

I'm defining my eslint rule using `TSESLint.RuleModule`, which types the context object as `TSESLint.RuleContext`, but this type doesn't include those comment functions. They are there, so I can cast `context` as `any` and call them, but this feels like a bug in the types.

**Comment by armano2** [comment]
@bschlenk 
```
const sourceCode = context.getSourceCode();
sourceCode.commentsExistBetween
```

**Comment by bschlenk** [comment]
Thanks armano2, that works! But now I'm hitting another issue. `getCommentsBefore` doesn't seem to work with TS nodes like `TSInterfaceDeclaration` or `TSTypeAliasDeclaration`. It always returns an empty array.
